quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index.rs

```diff
@@ -387,8 +400,93 @@ impl FileBackedIndex {
         Ok(true)
     }
 
-    /// Resets the checkpoint of a source. Returns whether a mutation occurred.
     pub(crate) fn reset_source_checkpoint(&mut self, source_id: &str) -> MetastoreResult<bool> {
         Ok(self.metadata.checkpoint.reset_source(source_id))
     }
+
+    pub(crate) fn create_delete_task(
+        &mut self,
+        delete_query: DeleteQuery,
+    ) -> MetastoreResult<DeleteTask> {
+        let now_timestamp = OffsetDateTime::now_utc().unix_timestamp();
+        let delete_task = DeleteTask {
+            create_timestamp: now_timestamp,
+            opstamp: self.stamper.stamp() as u64,
+            delete_query,
+        };
+        self.delete_tasks.push(delete_task.clone());
+        Ok(delete_task)
+    }
+
+    pub(crate) fn last_delete_opstamp(&self) -> u64 {
+        self.delete_tasks
+            .iter()
+            .map(|delete_task| delete_task.opstamp)
+            .max()
+            .unwrap_or(0)
+    }
+
+    pub(crate) fn delete_delete_tasks(&mut self) -> MetastoreResult<bool> {
+        self.delete_tasks.clear();
+        Ok(true)
+    }
+
+    pub(crate) fn update_splits_delete_opstamp(
+        &mut self,
+        split_ids: &[&str],
+        delete_opstamp: u64,
+    ) -> MetastoreResult<bool> {
+        for split_id in split_ids {
+            let split =
+                self.splits
+                    .get_mut(*split_id)
+                    .ok_or_else(|| MetastoreError::SplitsDoNotExist {
+                        split_ids: vec![split_id.to_string()],
+                    })?;
+            split.split_metadata.delete_opstamp = delete_opstamp;
+        }
+        Ok(true)
+    }
+
+    pub(crate) fn list_delete_tasks(&self, opstamp_start: u64) -> MetastoreResult<Vec<DeleteTask>> {
+        let delete_tasks = self
+            .delete_tasks
+            .iter()
+            .filter(|delete_task| delete_task.opstamp > opstamp_start)
+            .cloned()
+            .collect_vec();
+        Ok(delete_tasks)
+    }
+}
+
+/// Stamper provides Opstamps, which is just an auto-increment id to label
+/// a delete operation.
+/// Copy-pasted from tantivy with an `AtomicUsize` to have a broader support.
+#[derive(Clone, Default)]
+pub struct Stamper(Arc<AtomicUsize>);
```

Comment:
Interior mutability wiht an `Arc<AtomicUsize>` looks like a good idea but it is not here.

That stamper is owned by FileBackedIndex which is modified using a strange pattern to mimick the concept of transaction.
We acqurie a lock, mutate it, and if everything goes right, we store the modfiied copy.

By using this Arc<AtomicUsize> you are idea the fact that the FileBackedIndex ends up not being modified if a transaction fails. Just an usize would have been nicer.